File: spring-boot-s3-java-maven-spring-cloud-aws/src/main/java/io/pivotal/tsalm/s3/infrastructure/S3Service.java
package io.pivotal.tsalm.s3.infrastructure;

import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.util.IOUtils;
import org.springframework.core.io.Resource;
import org.springframework.core.io.ResourceLoader;
import org.springframework.core.io.WritableResource;
import org.springframework.stereotype.Service;

import java.io.IOException;
import java.io.OutputStream;
import java.util.UUID;

@Service
public class S3Service {

    private ResourceLoader resourceLoader;
    private AmazonS3 amazonS3;

    public S3Service(ResourceLoader resourceLoader, AmazonS3 amazonS3) {
        this.resourceLoader = resourceLoader;
        this.amazonS3 = amazonS3;
    }

    public Resource getObject(String bucketName, String key) throws S3KeyDoesNotExistException {
        final Resource resource = resourceLoader.getResource(String.format("s3://%s/%s", bucketName, key));
        if (resource.exists()) {
            return resource;
        }
        throw new S3KeyDoesNotExistException(bucketName, key);
    }

    public String putObject(String bucketName, Resource resource) throws IOException {
        String key = resource.getFilename() != null ? resource.getFilename() : UUID.randomUUID().toString();
        final WritableResource writableResource = (WritableResource) this.resourceLoader
                .getResource(String.format("s3://%s/%s", bucketName, key));

        try (OutputStream outputStream = writableResource.getOutputStream()) {
            IOUtils.copy(resource.getInputStream(), outputStream);
        }

        return key;
    }

    public void removeObject(String bucketName, String key) throws S3KeyDoesNotExistException {
        if (amazonS3.doesObjectExist(bucketName, key)) {
            amazonS3.deleteObject(bucketName, key);
        } else {
            throw new S3KeyDoesNotExistException(bucketName, key);
        }
    }
}

File: spring-boot-s3-kotlin-gradle-spring-cloud-aws/src/main/kotlin/io/pivotal/tsalm/s3/infrastructure/S3Service.kt
package io.pivotal.tsalm.s3.infrastructure

import com.amazonaws.services.s3.AmazonS3
import org.springframework.core.io.Resource
import org.springframework.core.io.ResourceLoader
import org.springframework.core.io.WritableResource
import org.springframework.stereotype.Repository
import java.util.*

@Repository
class S3Service(private val resourceLoader: ResourceLoader, private val amazonS3: AmazonS3) {

    @Throws(S3KeyDoesNotExistException::class)
    fun getObject(bucketName: String, key: String): Resource {
        val resource = resourceLoader.getResource("s3://$bucketName/$key")
        if (resource.exists()) {
            return resource
        }
        throw S3KeyDoesNotExistException(bucketName, key)
    }

    fun putObject(bucketName: String, resource: Resource): String {
        val key = resource.filename ?: UUID.randomUUID().toString()
        val writableResource = this.resourceLoader.getResource("s3://$bucketName/$key") as WritableResource
        writableResource.outputStream.use { fileOut ->
            resource.inputStream.copyTo(fileOut)
        }
        return key
    }

    @Throws(S3KeyDoesNotExistException::class)
    fun removeObject(bucketName: String, key: String) {
        if (amazonS3.doesObjectExist(bucketName, key)) {
            amazonS3.deleteObject(bucketName, key)
        } else {
            throw S3KeyDoesNotExistException(bucketName, key)
        }
    }
}